llama-index-core/llama_index/core/async_utils.py

```diff
@@ -21,17 +23,27 @@ def asyncio_module(show_progress: bool = False) -> Any:
 
 
 def asyncio_run(coro: Coroutine) -> Any:
-    """Gets an existing event loop to run the coroutine.
-
-    If there is no existing event loop, creates a new one.
+    """
+    Run coroutine synchronously.
     """
     try:
-        # this will fail if the event loop is already running
-        loop = asyncio.get_event_loop()
-        return loop.run_until_complete(coro)
+        _ = asyncio.get_running_loop()
     except RuntimeError:
         return asyncio.run(coro)
 
+    def main(q: SimpleQueue) -> None:
+        try:
+            q.put(asyncio.run(coro))
+        except BaseException as e:
+            q.put(e)
+
+    q: "SimpleQueue[Any]" = SimpleQueue()
+    t = Thread(target=main, args=(q,))
+    t.start(), t.join()
+    if isinstance(result := q.get(), BaseException):
+        raise result
+    return result
```

Comment:
I'm not 100% sure if I want to be spinning up threads. This can make debugging and error handling extremely difficult 

Generally if there is an error about nested async, that's more of an issue with a particular design, either in the framework (multiple nested async calls) or by the user 

Is there a specific case that was causing this? We should instead fix that code path if possible. 

Or if it's caused by a user, it can be fixed on their end either by using a more proper async setup, or using nest_asyncio (although avoiding that when possible is ideal)

Reply:
OK, I wasn't sure what the intent was, but as written, it fails when i tried to call `str(response)` when response is an AsyncStreamingResponse and when I have an active event loop (because I am running async).

Reply:
I guess the answer is then that we're not supposed to call `str` on `response` when running async.

Reply:
On the other hand, this `async_run` (with underscore) would then be equivalent to `asyncio.run` (with dot) and is not useful as a separate function.